File: src/main/scala/fitnesse/idea/lang/reference/FixtureMethodReferencesSearch.java
package fitnesse.idea.lang.reference;

import com.intellij.openapi.application.QueryExecutorBase;
import com.intellij.psi.PsiMethod;
import com.intellij.psi.PsiReference;
import com.intellij.psi.search.searches.MethodReferencesSearch;
import com.intellij.util.Processor;
import org.jetbrains.annotations.NotNull;

public class FixtureMethodReferencesSearch extends QueryExecutorBase<PsiReference, MethodReferencesSearch.SearchParameters> {
    public FixtureMethodReferencesSearch() {
        super(true);
    }

    @Override
    public void processQuery(@NotNull final MethodReferencesSearch.SearchParameters p, @NotNull final Processor<PsiReference> consumer) {
        final PsiMethod method = p.getMethod();

//        final PsiAnnotation stepAnnotation = CucumberJavaUtil.getCucumberStepAnnotation(method);
//        final String regexp = stepAnnotation != null ? CucumberJavaUtil.getPatternFromStepDefinition(stepAnnotation) : null;
//        if (regexp == null) {
//            return;
//        }
//
//        final String word = CucumberUtil.getTheBiggestWordToSearchByIndex(regexp);
//        if (StringUtil.isEmpty(word)) {
//            return;
//        }

        System.out.println("Search references for " + method);

//        SearchScope scope = p.getEffectiveSearchScope();
//        if (scope instanceof GlobalSearchScope) {
//            GlobalSearchScope restrictedScope = GlobalSearchScope.getScopeRestrictedByFileTypes((GlobalSearchScope) scope, FitnesseFileType.INSTANCE());
//            ReferencesSearch.search(new ReferencesSearch.SearchParameters(method, restrictedScope, false, p.getOptimizer())).forEach(consumer);
//        }
    }
}

File: src/main/scala/fitnesse/idea/fixturemethod/FixtureMethodReferencesSearch.scala
package fitnesse.idea.fixturemethod

import com.intellij.openapi.application.QueryExecutorBase
import com.intellij.openapi.project.Project
import com.intellij.psi.search.GlobalSearchScope
import com.intellij.psi.search.searches.MethodReferencesSearch
import com.intellij.psi.{PsiElement, PsiMethod, PsiReference}
import com.intellij.util.Processor
import fitnesse.idea.filetype.FitnesseFileType

import scala.collection.JavaConversions._

class FixtureMethodReferencesSearch extends QueryExecutorBase[PsiReference, MethodReferencesSearch.SearchParameters](true) {

  def processQuery(p: MethodReferencesSearch.SearchParameters, consumer: Processor[PsiReference]): Unit = {
    p.getMethod match {
      case method: PsiMethod =>
        p.getEffectiveSearchScope match {
          case scope: GlobalSearchScope =>
            val restrictedScope = GlobalSearchScope.getScopeRestrictedByFileTypes(scope, FitnesseFileType.INSTANCE)
            find(method.getName, method.getProject, restrictedScope)
              .map(fixtureMethod => consumer.process(new MethodReference(fixtureMethod) {
                override def resolve(): PsiElement = method
              }))
          case _ =>
        }
      case _ =>
    }
  }

  def find(key: String, project: Project, scope: GlobalSearchScope): List[FixtureMethod] = {
    val references: List[FixtureMethod] = FixtureMethodIndex.INSTANCE.get(key, project, scope).toList

    // TODO: Take into account the class name (via Stub structure?)
    references
  }

}
